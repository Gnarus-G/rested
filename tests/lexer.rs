use insta::assert_debug_snapshot;

macro_rules! assert_lexes {
    ($input:literal) => {
        let lexer = rested::lexer::Lexer::new($input);

        insta::with_settings!({
             description => $input
        }, {
            assert_debug_snapshot!(lexer.into_iter().collect::<Vec<rested::lexer::Token>>());
        })
    }
}

#[test]
fn lex_put_patch_delete() {
    assert_lexes!(
        r#"
put /api {}
patch /api {}
delete /api {}
"#
    );
}

#[test]
fn lex_string_literals() {
    assert_lexes!(r#""hello""#);

    assert_lexes!(r#""hello"#);

    assert_lexes!(
        r#"
"hello
"world
"#
    );

    assert_lexes!(r#" "" "" ``"#);

    assert_lexes!(r#" { "Bearer token" } "#);

    assert_lexes!(
        r#"`
{
    stuff
}`

`
stuff"#
    );
}

#[test]
fn lex_bools() {
    assert_lexes!("true false");
}

#[test]
fn lex_numbers() {
    assert_lexes!("123124 1.0");
}

#[test]
fn lex_get_url() {
    assert_lexes!("get http://localhost");
}

#[test]
fn lex_get_url_with_header() {
    assert_lexes!("get http://localhost { header \"Authorization\" \"Bearer token\" }");
}

#[test]
fn lex_get_url_over_many_lines() {
    assert_lexes!("get\nhttp://localhost");

    assert_lexes!(
        r#"get 
    http://localhost 
{
}"#
    );
}

#[test]
fn lex_get_url_with_header_and_body() {
    assert_lexes!(
        r#"
post http://localhost { 
    header "Authorization" "Bearer token" 
    body "{neet: 1337}" 
}"#
    );
}

#[test]
fn lex_call_expression() {
    assert_lexes!(r#"env() env("stuff")"#);
}

#[test]
fn lex_template_literals() {
    assert_lexes!(r#"`stuff${"interpolated"}(things${env("dead_night")}` `dohickeys`"#);

    assert_lexes!(r#"`a${"temp"}` }}"#);

    assert_lexes!(
        r#"
let b = `asdf ${}`
"#
    );

    assert_lexes!(r#"` ${} ${} `"#);
    assert_lexes!(r#"`head ${} ${} tail`"#);

    assert_lexes!(
        r#"
let b = `asdf ${test} ${} ${"word"}`
"#
    );

    assert_lexes!(
        r#"
let b = `asdf ${}` 
let c = {}
"#
    );

    assert_lexes!(r#"`${test} ${df} `"#);
    assert_lexes!(r#"`asdf ${test} ${df} `"#);
}

#[test]
fn lex_json_object() {
    assert_lexes!(
        r#"
let obj = {
    key: "value",
    otherKey: ["val", "vlue"]
}
"#
    );
}
