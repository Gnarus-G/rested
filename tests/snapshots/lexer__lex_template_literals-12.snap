---
source: tests/lexer.rs
description: "`string${env(\"base\")}morestring${true}evenmore${\"expr string\"}`"
expression: "lexer.into_iter().collect::<Vec<rested::lexer::Token>>()"
---
[
    OpeningBackTick("`") at Position { value: 0, line: 0, col: 0 },
    StringLiteral("string") at Position { value: 1, line: 0, col: 1 },
    DollarSignLBracket("${") at Position { value: 7, line: 0, col: 7 },
    Ident("env") at Position { value: 9, line: 0, col: 9 },
    LParen("(") at Position { value: 12, line: 0, col: 12 },
    StringLiteral("\"base\"") at Position { value: 13, line: 0, col: 13 },
    RParen(")") at Position { value: 19, line: 0, col: 19 },
    RBracket("}") at Position { value: 20, line: 0, col: 20 },
    StringLiteral("morestring") at Position { value: 21, line: 0, col: 21 },
    DollarSignLBracket("${") at Position { value: 31, line: 0, col: 31 },
    Boolean("true") at Position { value: 33, line: 0, col: 33 },
    RBracket("}") at Position { value: 37, line: 0, col: 37 },
    StringLiteral("evenmore") at Position { value: 38, line: 0, col: 38 },
    DollarSignLBracket("${") at Position { value: 46, line: 0, col: 46 },
    StringLiteral("\"expr string\"") at Position { value: 48, line: 0, col: 48 },
    RBracket("}") at Position { value: 61, line: 0, col: 61 },
    ClosingBackTick("`") at Position { value: 62, line: 0, col: 62 },
]
